src/dbnode/encoding/series_iterator_accumulator.go

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/x/ident"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+var _ SeriesIteratorAccumulator = (*seriesIteratorAccumulator)(nil)
+
+type seriesIteratorAccumulator struct {
+	id              ident.ID
+	nsID            ident.ID
+	start           time.Time
+	end             time.Time
+	iters           iterators
+	tagIterator     ident.TagIterator
+	seriesIterators []SeriesIterator
+	err             error
+	firstNext       bool
+	closed          bool
+}
+
+// SeriesAccumulatorOptions are options for a SeriesIteratorAccumulator.
+type SeriesAccumulatorOptions struct {
+	// RetainTags determines if tags should be preserved after the accumulator is
+	// exhausted. If set to true, the accumulator retains a copy of the tags.
+	RetainTags bool
```

Comment:
Could we use Rewind for this case?

Just for helping me review the PR, when do we need to use RetainTags?

Reply:
So it's a bit weird; since `Accumulator` uses the base `encoding.interface`, as part of its iteration, it closes exhausted sub-iterators which frees their tag iterators, too. Functionally, this means that if you call `Tags()` on the Accumulator after iterating through the values, you wouldn't be able to get any Tags back, since it wouldn't have any more data.